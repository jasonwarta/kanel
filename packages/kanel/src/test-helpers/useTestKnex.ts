import knex, { type Knex } from "knex";
import { afterAll, beforeAll } from "vitest";

import usePostgresContainer from "./usePostgresContainer";

type Connection = {
  host: string;
  port: number;
  password: string;
  user: string;
  database: string;
};

const useTestKnex = (): readonly [
  () => Knex<any, any[]>,
  string,
  () => Connection,
] => {
  let knexInstance: Knex;
  const databaseName = `test_${Math.ceil(Math.random() * 1000)}`;
  let connection: Connection;

  const getContainer = usePostgresContainer();

  beforeAll(async () => {
    const container = getContainer();
    connection = {
      host: container.getHost(),
      port: container.getMappedPort(5432),
      password: "postgres",
      user: "postgres",
      database: "postgres",
    };

    const setupKnexInstance = knex({
      client: "postgres",
      connection,
    });
    await setupKnexInstance.raw("create database ??", [databaseName]);
    await setupKnexInstance.destroy();

    connection.database = databaseName;
    knexInstance = knex({
      client: "postgres",
      connection,
    });
  });

  afterAll(async () => {
    const container = getContainer();
    const connection = {
      host: container.getHost(),
      port: container.getMappedPort(5432),
      password: "postgres",
      user: "postgres",
    };

    const setupKnexInstance = knex({
      client: "postgres",
      connection: { ...connection, database: "postgres" },
    });

    setupKnexInstance
      .raw(`drop database ${databaseName} with (force)`)
      .then(() => setupKnexInstance.destroy());

    await knexInstance.destroy();
  });

  return [() => knexInstance, databaseName, () => connection] as const;
};

export default useTestKnex;
