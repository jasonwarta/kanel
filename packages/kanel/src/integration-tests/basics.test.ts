import { describe, expect, it, vi } from "vitest";

import processDatabase from "../processDatabase";
import useTestKnex from "../test-helpers/useTestKnex";
import useSchema from "../test-helpers/useSchema";

vi.mock("../writeFile", () => ({
  default: vi.fn(),
}));

import writeFile from "../writeFile";
const mockedWriteFile = vi.mocked(writeFile);

function getResults(): { [fullPath: string]: string[] } {
  return mockedWriteFile.mock.calls.reduce(
    (acc, [args]) => {
      acc[args.fullPath] = args.lines;
      return acc;
    },
    {} as Record<string, string[]>,
  );
}

describe("Basic integration tests", () => {
  const [getKnex, _, getConnection] = useTestKnex();
  useSchema(getKnex, "test");

  it("should process a simple table", async () => {
    const db = getKnex();
    await db.raw("create table test.some_table (id integer)");

    await processDatabase({ connection: getConnection() });

    const result = getResults();
    expect(result).toMatchSnapshot();
  });

  // Database with a couple of tables with FK relations, enums, views, etc.
  it("should process the fundamentals", async () => {
    const db = getKnex();
    const sql = `
      create table test.some_table (id integer);
      create table test.some_other_table (id integer primary key);
      alter table test.some_table add column some_other_table_id integer references test.some_other_table (id);
      create type test.some_enum as enum ('value1', 'value2');
      create view test.some_view as select * from test.some_table;
      create materialized view test.some_materialized_view as select * from test.some_table;
      create type test.some_range as range (subtype = test.some_enum);
      create domain test.some_domain as text;
      create function test.some_function() returns void language sql as $$ $$;
      create procedure test.some_procedure() language sql as $$ $$;
    `;
    await db.raw(sql);

    await processDatabase({ connection: getConnection() });

    const result = getResults();
    expect(result).toMatchSnapshot();
  });
});
