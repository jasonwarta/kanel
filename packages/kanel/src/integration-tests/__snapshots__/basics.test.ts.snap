// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Basic integration tests > should process a simple table 1`] = `
{
  "test/SomeTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the table test.some_table */",
    "export default interface SomeTable {",
    "  id: number | null;",
    "}",
    "",
    "/** Represents the initializer for the table test.some_table */",
    "export interface SomeTableInitializer {",
    "  id?: number | null;",
    "}",
    "",
    "/** Represents the mutator for the table test.some_table */",
    "export interface SomeTableMutator {",
    "  id?: number | null;",
    "}",
    "",
  ],
}
`;

exports[`Basic integration tests > should process the fundamentals 1`] = `
{
  "test/SomeDomain.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the domain test.some_domain */",
    "type SomeDomain = string;",
    "",
    "export default SomeDomain;",
    "",
  ],
  "test/SomeEnum.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the enum test.some_enum */",
    "enum SomeEnum {",
    "  value1 = 'value1',",
    "  value2 = 'value2',",
    "};",
    "",
    "export default SomeEnum;",
    "",
  ],
  "test/SomeMaterializedView.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the materializedView test.some_materialized_view */",
    "export default interface SomeMaterializedView {",
    "  id: number | null;",
    "",
    "  some_other_table_id: number | null;",
    "}",
    "",
  ],
  "test/SomeOtherTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Identifier type for test.some_other_table */",
    "export type SomeOtherTableId = number & { __brand: 'test.some_other_table' };",
    "",
    "/** Represents the table test.some_other_table */",
    "export default interface SomeOtherTable {",
    "  id: SomeOtherTableId;",
    "}",
    "",
    "/** Represents the initializer for the table test.some_other_table */",
    "export interface SomeOtherTableInitializer {",
    "  id: SomeOtherTableId;",
    "}",
    "",
    "/** Represents the mutator for the table test.some_other_table */",
    "export interface SomeOtherTableMutator {",
    "  id?: SomeOtherTableId;",
    "}",
    "",
  ],
  "test/SomeRange.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the range test.some_range */",
    "type SomeRange = string;",
    "",
    "export default SomeRange;",
    "",
  ],
  "test/SomeTable.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "import type { SomeOtherTableId } from './SomeOtherTable';",
    "",
    "/** Represents the table test.some_table */",
    "export default interface SomeTable {",
    "  id: number | null;",
    "",
    "  some_other_table_id: SomeOtherTableId | null;",
    "}",
    "",
    "/** Represents the initializer for the table test.some_table */",
    "export interface SomeTableInitializer {",
    "  id?: number | null;",
    "",
    "  some_other_table_id?: SomeOtherTableId | null;",
    "}",
    "",
    "/** Represents the mutator for the table test.some_table */",
    "export interface SomeTableMutator {",
    "  id?: number | null;",
    "",
    "  some_other_table_id?: SomeOtherTableId | null;",
    "}",
    "",
  ],
  "test/SomeView.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "/** Represents the view test.some_view */",
    "export default interface SomeView {",
    "  id: number;",
    "",
    "  some_other_table_id: number;",
    "}",
    "",
  ],
  "test/some_function.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "export interface some_function_params {",
    "}",
    "",
  ],
  "test/some_procedure.ts": [
    "// @generated",
    "// This file is automatically generated by Kanel. Do not modify manually.",
    "",
    "export interface some_procedure_params {",
    "}",
    "",
  ],
}
`;
